File: src/engine/model/EntryPoint.cpp
/*
 * EntryPoint.cpp
 *
 *  Created on: Mar 5, 2014
 *      Author: Stephan Opfer
 */

#include "engine/model/EntryPoint.h"
#include "engine/model/Transition.h"
#include "engine/model/State.h"
#include "engine/model/Task.h"

namespace alica
{

	EntryPoint::EntryPoint()
	{
	}

	EntryPoint::~EntryPoint()
	{
	}

	void EntryPoint::computeReachabilitySet()
	{
		list<State*> queue;
		queue.push_front(this->state);
		State* cs;
		while (queue.size() > 0)
		{
			cs = *queue.begin();
			queue.pop_front();
			if (this->reachableStates.insert(cs).second)
			{
				for (Transition* t : cs->getOutTransitions())
				{
					queue.push_back(t->getOutState());
				}
			}

		}
	}

	string EntryPoint::toString()
	{
		stringstream ss;
		ss << "#EntryPoint: " << this->name << " " << this->id << endl;
		ss << "\t MinCardinality: " << this->minCardinality << endl;
		ss << "\t MaxCardinality: " << this->maxCardinality << endl;
		ss << "\t Task:" << endl;
		if (this->task != NULL)
		{
			ss << "\t" << this->task->getId() << " " << this->task->getName();
		}
		else
		{
			ss << "null";
		}
		ss << endl;
		ss << "\t Initial State:" << endl;
		if (this->state != NULL)
		{
			ss << "\t" << this->state->getId() << " " << this->state->getName();
		}
		else
		{
			ss << "null";
		}
		ss << endl;
		ss << "#EndEntryPoint" << endl;
		return ss.str();

	}

	bool EntryPoint::compareTo(const EntryPoint* ep1, const EntryPoint* ep2)
	{
		return (ep1->getTask()->getId() < ep2->getTask()->getId());
	}

//================== Getter and Setter =============

	const Task* EntryPoint::getTask() const
	{
		return task;
	}

	void EntryPoint::setTask(const Task* task)
	{
		this->task = task;
	}

	void EntryPoint::setPlan(Plan* plan)
	{
		this->plan = plan;
	}

	Plan* EntryPoint::getPlan() const
	{
		return plan;
	}

	const int EntryPoint::getMaxCardinality() const
	{
		return this->maxCardinality;
	}

	void EntryPoint::setMaxCardinality(int maxCardinality)
	{
		this->maxCardinality = maxCardinality;
	}

	const int EntryPoint::getMinCardinality() const
	{
		return this->minCardinality;
	}

	void EntryPoint::setMinCardinality(int minCardinality)
	{
		this->minCardinality = minCardinality;
	}

	void EntryPoint::setSuccessRequired(bool successRequired)
	{
		this->successRequired = successRequired;
	}

	const bool EntryPoint::getSuccessRequired() const
	{
		return this->successRequired;
	}
	const unordered_set<State*>& EntryPoint::getReachableStates() const
	{
		return reachableStates;
	}

	const State* EntryPoint::getState() const
	{
		return state;
	}

	void EntryPoint::setState(State* state)
	{
		this->state = state;
	}

	void EntryPoint::setReachableStates(const unordered_set<State*>& reachableStates)
	{
		this->reachableStates = reachableStates;
	}

	bool EntryPoint::isSuccessRequired() const
	{
		return successRequired;
	}

} /* namespace Alica */

File: supplementary/alica_codegeneration/alica-plan-designer-fx-model/src/main/java/de/unikassel/vs/alica/planDesigner/alicamodel/EntryPoint.java
package de.unikassel.vs.alica.planDesigner.alicamodel;

import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleIntegerProperty;
import javafx.beans.property.SimpleObjectProperty;

public class EntryPoint extends PlanElement {

    protected final SimpleBooleanProperty successRequired = new SimpleBooleanProperty(this, "successRequired", false);
    protected final SimpleIntegerProperty minCardinality = new SimpleIntegerProperty(this, "minCardinality", 0);
    protected final SimpleIntegerProperty maxCardinality = new SimpleIntegerProperty(this, "maxCardinality", 2147483647);
    protected final SimpleObjectProperty<Task> task = new SimpleObjectProperty<>(this, "task", null);
    protected final SimpleObjectProperty<State> state = new SimpleObjectProperty<>(this, "state", null);
    protected final SimpleObjectProperty<Plan> plan = new SimpleObjectProperty<>(this, "plan", null);


    public boolean getSuccessRequired() {
        return successRequired.get();
    }
    public void setSuccessRequired(boolean successRequired) {
        this.successRequired.set(successRequired);
    }
    public SimpleBooleanProperty successRequiredProperty() {
        return successRequired;
    }

    public int getMinCardinality() {
        return minCardinality.get();
    }
    public void setMinCardinality(int minCardinality) {
        this.minCardinality.set(minCardinality);
    }
    public SimpleIntegerProperty minCardinalityProperty() {
        return minCardinality;
    }

    public int getMaxCardinality() {
        return maxCardinality.get();
    }
    public void setMaxCardinality(int maxCardinality) {
        this.maxCardinality.set(maxCardinality);
    }
    public SimpleIntegerProperty maxCardinalityProperty() {
        return maxCardinality;
    }

    public Task getTask() {
        return task.get();
    }
    public void setTask(Task task) {
        this.task.set(task);
    }
    public SimpleObjectProperty<Task> taskProperty() {
        return task;
    }

    public State getState() {
        return state.get();
    }
    public void setState(State state) {
        this.state.set(state);
    }
    public SimpleObjectProperty<State> stateProperty() {
        return state;
    }

    public Plan getPlan() {
        return plan.get();
    }
    public void setPlan(Plan plan) {
        this.plan.set(plan);
    }
    public SimpleObjectProperty<Plan> planProperty() {
        return plan;
    }

    public void registerDirtyFlag(ChangeListenerForDirtyFlag listener)
    {
        this.successRequired.addListener(listener);
        this.minCardinality.addListener(listener);
        this.maxCardinality.addListener(listener);
        this.task.addListener(listener);
        this.state.addListener(listener);
        this.plan.addListener(listener);
    }
}
